### PR TITLE
Add support to expose custom ports

### PR DESCRIPTION
**Pull request by luizm**
I'd like to expose other ports in my spark app, similar to what the spark-operator already has done to prometheus.

Its related to: 

https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/1202 
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/1120

### CODE REVIEW

**Comment by luizm** [comment]
@googlebot I signed it!

File: pkg/webhook/patch.go [hunk 1]
```diff
@@ -384,13 +385,32 @@ func getPrometheusConfigPatches(pod *corev1.Pod, app *v1beta2.SparkApplication)
...
+		glog.Warningf("could not expose port %d to scrape metrics outside the pod", promPort)
 		return nil
 	}
-	patchOps = append(patchOps, *portPatchOp)
+	patchOps = append(patchOps, *promPortPatchOp)
+	return patchOps
+}
+
+func addContainerPorts(pod *corev1.Pod, app *v1beta2.SparkApplication) []patchOperation {
+	var PortMaps []v1beta2.Port
```
Comment (liyinan926) [inline pkg/webhook/patch.go:398]: Local variables should start with a lower-case letter. Can you rename this to `ports`?

File: pkg/webhook/patch.go [hunk 2]
```diff
@@ -384,13 +385,32 @@ func getPrometheusConfigPatches(pod *corev1.Pod, app *v1beta2.SparkApplication)
...
+		PortMaps = app.Spec.Driver.Ports
+	} else if util.IsExecutorPod(pod) {
+		PortMaps = app.Spec.Executor.Ports
+	}
+
+	var patchOps []patchOperation
+	for _, p := range PortMaps {
+		portPatchOp := addContainerPort(pod, p.ContainerPort, p.Protocol, p.Name)
+		if portPatchOp == nil {
+			return nil
```
Comment (liyinan926) [inline pkg/webhook/patch.go:410]: Should this be `continue`?

Comment (luizm) [reply pkg/webhook/patch.go:410]: yep.. thank you

File: pkg/webhook/patch.go [hunk 3]
```diff
@@ -384,13 +385,32 @@ func getPrometheusConfigPatches(pod *corev1.Pod, app *v1beta2.SparkApplication)
...
+		glog.Warningf("could not expose port %d to scrape metrics outside the pod", promPort)
 		return nil
 	}
-	patchOps = append(patchOps, *portPatchOp)
+	patchOps = append(patchOps, *promPortPatchOp)
+	return patchOps
+}
+
+func addContainerPorts(pod *corev1.Pod, app *v1beta2.SparkApplication) []patchOperation {
+	var portMaps []v1beta2.Port
```
Comment (liyinan926) [inline pkg/webhook/patch.go:398]: Nit: the name suggests this is a map, although it is not.

File: pkg/webhook/patch.go [hunk 4]
```diff
@@ -384,13 +385,32 @@ func getPrometheusConfigPatches(pod *corev1.Pod, app *v1beta2.SparkApplication)
...
+		portMaps = app.Spec.Driver.Ports
+	} else if util.IsExecutorPod(pod) {
+		portMaps = app.Spec.Executor.Ports
+	}
 
+	var patchOps []patchOperation
+	for _, p := range portMaps {
+		portPatchOp := addContainerPort(pod, p.ContainerPort, p.Protocol, p.Name)
+		if portPatchOp == nil {
+			glog.Warningf("could not expose port named %s", p.Name)
```
Comment (liyinan926) [inline pkg/webhook/patch.go:410]: You need to add a `continue` after the log.

**Review by liyinan926 — APPROVED** [review]
LGTM. Thanks!

**Comment by luizm** [comment]
@liyinan926 the docker image needs to be updated manually? 

https://console.cloud.google.com/gcr/images/spark-operator/GLOBAL/spark-operator?gcrImageListsize=30&gcrImageListsort=-uploaded

**Pull request merged**